modules/nlu/src/backend/tools/strings.ts

```diff
@@ -0,0 +1,78 @@
+import _ from 'lodash'
+
+/** Splits a string in all its N-grams */
+export function ngram(value: string, n: number): string[] {
+  const nGrams: string[] = []
+  let index = value.length - n + 1
+
+  while (index--) {
+    nGrams.push(value.slice(index, index + n))
+  }
+
+  return nGrams
+}
+
+export function vocabNGram(tokens: string[]): string[] {
+  const plainTokens = tokens
+    .map(x => x.replace('_', '\u2581')) // We want to discover real language-specific chars
```

Comment:
I don't get why you replace _ with sentencepiece whitespace char ?

Reply:
oups, meant to replace \u2581 by empty string (to get rid of them), just in case any were passed in